# LAB | Ensemble Methods

**Load the data**

In this challenge, we will be working with the same Spaceship Titanic data, like the previous Lab. The data can be found here:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv

Metadata

https://github.com/data-bootcamp-v4/data/blob/main/spaceship_titanic.md

In this Lab, you should try different ensemble methods in order to see if can obtain a better model than before. In order to do a fair comparison, you should perform the same feature scaling, engineering applied in previous Lab.

In [1]:
#Libraries
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split

from sklearn.datasets import  fetch_california_housing
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import BaggingRegressor, RandomForestRegressor,AdaBoostRegressor, GradientBoostingRegressor

from sklearn.preprocessing import MinMaxScaler, StandardScaler
from sklearn.metrics import r2_score, mean_absolute_error, mean_squared_error, root_mean_squared_error

In [2]:
spaceship = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv")
spaceship.head()

PassengerId HomePlanet CryoSleep  Cabin  Destination   Age    VIP  \
0     0001_01     Europa     False  B/0/P  TRAPPIST-1e  39.0  False   
1     0002_01      Earth     False  F/0/S  TRAPPIST-1e  24.0  False   
2     0003_01     Europa     False  A/0/S  TRAPPIST-1e  58.0   True   
3     0003_02     Europa     False  A/0/S  TRAPPIST-1e  33.0  False   
4     0004_01      Earth     False  F/1/S  TRAPPIST-1e  16.0  False   

   RoomService  FoodCourt  ShoppingMall     Spa  VRDeck               Name  \
0          0.0        0.0           0.0     0.0     0.0    Maham Ofracculy   
1        109.0        9.0          25.0   549.0    44.0       Juanna Vines   
2         43.0     3576.0           0.0  6715.0    49.0      Altark Susent   
3          0.0     1283.0         371.0  3329.0   193.0       Solam Susent   
4        303.0       70.0         151.0   565.0     2.0  Willy Santantines   

   Transported  
0        False  
1         True  
2        False  
3        False  
4         True

Now perform the same as before:
- Feature Scaling
- Feature Selection


In [3]:
spaceship.dropna(inplace = True)
spaceship['Cabin'] = spaceship['Cabin'].str.split('/').str[0]
spaceship = spaceship.drop(columns = ['PassengerId', 'Name'])
spaceship = pd.get_dummies(spaceship)
# sp_features = spaceship.select_dtypes (include = 'number')
# sp_targets = spaceship.select_dtypes (exclude = 'number')
sp_targets = spaceship['Transported']
sp_features = spaceship.drop(columns=['Transported'])

In [23]:
sp_features

Age  RoomService  FoodCourt  ShoppingMall     Spa  VRDeck  \
0     39.0          0.0        0.0           0.0     0.0     0.0   
1     24.0        109.0        9.0          25.0   549.0    44.0   
2     58.0         43.0     3576.0           0.0  6715.0    49.0   
3     33.0          0.0     1283.0         371.0  3329.0   193.0   
4     16.0        303.0       70.0         151.0   565.0     2.0   
...    ...          ...        ...           ...     ...     ...   
8688  41.0          0.0     6819.0           0.0  1643.0    74.0   
8689  18.0          0.0        0.0           0.0     0.0     0.0   
8690  26.0          0.0        0.0        1872.0     1.0     0.0   
8691  32.0          0.0     1049.0           0.0   353.0  3235.0   
8692  44.0        126.0     4688.0           0.0     0.0    12.0   

      HomePlanet_Earth  HomePlanet_Europa  HomePlanet_Mars  CryoSleep_False  \
0                False               True            False             True   
1                 True              False            False             True   
2                False               True            False             True   
3                False               True            False             True   
4                 True              False            False             True   
...                ...                ...              ...              ...   
8688             False               True            False             True   
8689              True              False            False            False   
8690              True              False            False             True   
8691             False               True            False             True   
8692             False               True            False             True   

      ...  Cabin_D  Cabin_E  Cabin_F  Cabin_G  Cabin_T  \
0     ...    False    False    False    False    False   
1     ...    False    False     True    False    False   
2     ...    False    False    False    False    False   
3     ...    False    False    False    False    False   
4     ...    False    False     True    False    False   
...   ...      ...      ...      ...      ...      ...   
8688  ...    False    False    False    False    False   
8689  ...    False    False    False     True    False   
8690  ...    False    False    False     True    False   
8691  ...    False     True    False    False    False   
8692  ...    False     True    False    False    False   

      Destination_55 Cancri e  Destination_PSO J318.5-22  \
0                       False                      False   
1                       False                      False   
2                       False                      False   
3                       False                      False   
4                       False                      False   
...                       ...                        ...   
8688                     True                      False   
8689                    False                       True   
8690                    False                      False   
8691                     True                      False   
8692                    False                      False   

      Destination_TRAPPIST-1e  VIP_False  VIP_True  
0                        True       True     False  
1                        True       True     False  
2                        True      False      True  
3                        True       True     False  
4                        True       True     False  
...                       ...        ...       ...  
8688                    False      False      True  
8689                    False       True     False  
8690                     True       True     False  
8691                    False       True     False  
8692                     True       True     False  

[6606 rows x 24 columns]

**Perform Train Test Split**

In [4]:
x_train, x_test, y_train, y_test = train_test_split (sp_features, sp_targets,test_size = 0.2, random_state=0)
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors = 3)
from sklearn.impute import SimpleImputer
imputer = SimpleImputer(strategy='median')
x_train_clean = imputer.fit_transform(x_train)
x_test_clean = imputer.transform(x_test)
knn.fit(x_train_clean, y_train)
y_pred = knn.predict(x_test_clean)
evaluation = knn.score(x_test_clean, y_test)*100
evaluation

C:\Users\a_gal\anaconda3\Lib\site-packages\joblib\externals\loky\backend\context.py:136: UserWarning: Could not find the number of physical cores for the following reason:
[WinError 2] The system cannot find the file specified
Returning the number of logical cores instead. You can silence this warning by setting LOKY_MAX_CPU_COUNT to the number of cores you want to use.
  warnings.warn(
  File "C:\Users\a_gal\anaconda3\Lib\site-packages\joblib\externals\loky\backend\context.py", line 257, in _count_physical_cores
    cpu_info = subprocess.run(
        "wmic CPU Get NumberOfCores /Format:csv".split(),
        capture_output=True,
        text=True,
    )
  File "C:\Users\a_gal\anaconda3\Lib\subprocess.py", line 554, in run
    with Popen(*popenargs, **kwargs) as process:
         ~~~~~^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\a_gal\anaconda3\Lib\subprocess.py", line 1039, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
    ~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^

75.642965204236

**Model Selection** - now you will try to apply different ensemble methods in order to get a better model

- Bagging and Pasting

In [5]:
from sklearn.preprocessing import MinMaxScaler
normalizer = MinMaxScaler()
normalizer.fit(x_train)
X_train_norm_np = normalizer.transform(x_train)
X_test_norm_np = normalizer.transform(x_test)

In [6]:
X_train_norm_df = pd.DataFrame(X_train_norm_np, columns = x_train.columns, index=x_train.index)
X_test_norm_df = pd.DataFrame(X_test_norm_np, columns = x_test.columns, index=x_test.index)
X_train_norm_df.head()

Age  RoomService  FoodCourt  ShoppingMall       Spa    VRDeck  \
3432  0.405063      0.00000   0.000000      0.000000  0.000000  0.000000   
7312  0.050633      0.00000   0.000000      0.000000  0.000000  0.000000   
2042  0.379747      0.00000   0.007916      0.000000  0.051276  0.000000   
4999  0.215190      0.00131   0.000000      0.046111  0.016378  0.000049   
5755  0.329114      0.00000   0.000000      0.000000  0.000000  0.000000   

      HomePlanet_Earth  HomePlanet_Europa  HomePlanet_Mars  CryoSleep_False  \
3432               0.0                0.0              1.0              0.0   
7312               1.0                0.0              0.0              0.0   
2042               1.0                0.0              0.0              1.0   
4999               0.0                0.0              1.0              1.0   
5755               1.0                0.0              0.0              0.0   

      ...  Cabin_D  Cabin_E  Cabin_F  Cabin_G  Cabin_T  \
3432  ...      0.0      1.0      0.0      0.0      0.0   
7312  ...      0.0      0.0      0.0      1.0      0.0   
2042  ...      0.0      0.0      1.0      0.0      0.0   
4999  ...      0.0      1.0      0.0      0.0      0.0   
5755  ...      0.0      0.0      0.0      1.0      0.0   

      Destination_55 Cancri e  Destination_PSO J318.5-22  \
3432                      0.0                        0.0   
7312                      0.0                        0.0   
2042                      0.0                        0.0   
4999                      0.0                        0.0   
5755                      0.0                        0.0   

      Destination_TRAPPIST-1e  VIP_False  VIP_True  
3432                      1.0        1.0       0.0  
7312                      1.0        1.0       0.0  
2042                      1.0        1.0       0.0  
4999                      1.0        1.0       0.0  
5755                      1.0        1.0       0.0  

[5 rows x 24 columns]

In [14]:
from sklearn.ensemble import BaggingClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score

bagging_clf = BaggingClassifier(
    estimator=DecisionTreeClassifier(max_depth=20),
    n_estimators=100,
    max_samples=1000
)

bagging_clf.fit(X_train_norm_df, y_train)
y_pred = bagging_clf.predict(X_test_norm_df)

# Now .score() returns Accuracy (e.g., 0.78)
print(f"Accuracy: {bagging_clf.score(X_test_norm_df, y_test) * 100:.2f}%")

Accuracy: 78.06%


- Random Forests

In [15]:
from sklearn.ensemble import RandomForestClassifier

forest = RandomForestClassifier(n_estimators=100, max_depth=20, random_state=0)
forest.fit(X_train_norm_df, y_train)

print(f"Accuracy: {forest.score(X_test_norm_df, y_test) * 100:.2f}%")

Accuracy: 79.58%


- Gradient Boosting

In [16]:
from sklearn.ensemble import AdaBoostClassifier

ada_clf = AdaBoostClassifier(
    estimator=DecisionTreeClassifier(max_depth=1), # AdaBoost usually uses depth=1 (Stumps)
    n_estimators=100
)
ada_clf.fit(X_train_norm_df, y_train)

print(f"Accuracy: {ada_clf.score(X_test_norm_df, y_test) * 100:.2f}%")

Accuracy: 78.14%


- Adaptive Boosting

In [17]:
from sklearn.ensemble import GradientBoostingClassifier

gb_clf = GradientBoostingClassifier(max_depth=5, n_estimators=100)
gb_clf.fit(X_train_norm_df, y_train)

print(f"Accuracy: {gb_clf.score(X_test_norm_df, y_test) * 100:.2f}%")

Accuracy: 78.59%


Which model is the best and why?

The jump from 75% (KNN) to ~79.5% (Random Forest) is actually a huge improvement but the fact that Bagging, AdaBoost, Gradient Boosting, and Random Forest are all stuck between 78% and 79.6% suggests that you have hit the limit of your current data, further analysis should be made to understand which one is really better model.